File: golang/transport.go
// FILE: golang/transport.go
// PURPOSE: Own the live WebSocket peer transport used by caller handles and service runners.
// OWNS: Dial/register/read/write/pending-request routing for v5 peers.
// EXPORTS: none.
// DOCS: docs/spec.md, agent_chat/go_v5_api_surface_2026-03-25.md
package multifrost

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"sync/atomic"

	"github.com/coder/websocket"
)

type transportReply struct {
	frame Frame
	err   error
}

type peerTransport struct {
	conn       *websocket.Conn
	peerID     string
	class      PeerClass
	writeMu    sync.Mutex
	pendingMu  sync.Mutex
	pending    map[string]chan transportReply
	dispatch   func(Frame)
	closed     chan struct{}
	closeErr   atomic.Value
	closedOnce sync.Once
}

func dialPeerTransport(
	ctx context.Context,
	endpoint string,
	peerID string,
	class PeerClass,
	dispatch func(Frame),
) (*peerTransport, error) {
	if ctx == nil {
		ctx = context.Background()
	}

	conn, _, err := websocket.Dial(ctx, endpoint, &websocket.DialOptions{})
	if err != nil {
		return nil, &TransportError{Origin: ErrorOriginTransport, Err: err}
	}

	transport := &peerTransport{
		conn:     conn,
		peerID:   peerID,
		class:    class,
		pending:  make(map[string]chan transportReply),
		dispatch: dispatch,
		closed:   make(chan struct{}),
	}

	if err := transport.register(ctx); err != nil {
		_ = conn.Close(websocket.StatusInternalError, "")
		return nil, err
	}

	go transport.readLoop()
	return transport, nil
}

func (t *peerTransport) register(ctx context.Context) error {
	body := RegisterBody{
		PeerID: t.peerID,
		Class:  t.class,
	}
	bodyBytes, err := EncodeMsgpack(body)
	if err != nil {
		return &TransportError{Origin: ErrorOriginTransport, Err: err}
	}

	envelope := NewEnvelope(KindRegister, t.peerID, RouterPeerID)
	frameBytes, err := EncodeFrame(envelope, bodyBytes)
	if err != nil {
		return err
	}

	if err := t.writeBinary(ctx, frameBytes); err != nil {
		return err
	}

	msgType, bytes, err := t.conn.Read(ctx)
	if err != nil {
		return t.wrapReadError(err)
	}
	if msgType != websocket.MessageBinary {
		return &RouterError{
			Origin:  ErrorOriginProtocol,
			Code:    "non_binary_message",
			Message: "router must send binary websocket frames",
		}
	}

	frame, err := DecodeFrame(bytes)
	if err != nil {
		return &RouterError{
			Origin:  ErrorOriginProtocol,
			Code:    "malformed_frame",
			Message: err.Error(),
		}
	}

	switch frame.Envelope.Kind {
	case KindResponse:
		var ack RegisterAckBody
		if err := DecodeMsgpack(frame.BodyBytes, &ack); err != nil {
			return &RegistrationError{
				Origin: ErrorOriginRegistration,
				Err:    err,
			}
		}
		if !ack.Accepted {
			reason := "router rejected registration"
			if ack.Reason != nil && *ack.Reason != "" {
				reason = *ack.Reason
			}
			return &RegistrationError{
				Origin: ErrorOriginRegistration,
				Reason: reason,
			}
		}
		return nil
	case KindError:
		var body ErrorBody
		if err := DecodeMsgpack(frame.BodyBytes, &body); err != nil {
			return &RegistrationError{
				Origin: ErrorOriginRegistration,
				Err:    err,
			}
		}
		return &RegistrationError{
			Origin: ErrorOriginRegistration,
			Reason: body.Message,
		}
	default:
		return &RegistrationError{
			Origin: ErrorOriginRegistration,
			Reason: fmt.Sprintf("unexpected register reply kind: %s", frame.Envelope.Kind),
		}
	}
}

func (t *peerTransport) request(ctx context.Context, envelope Envelope, bodyBytes []byte) (Frame, error) {
	if t.isClosed() {
		return Frame{}, t.transportClosedError()
	}

	replyCh := make(chan transportReply, 1)
	t.pendingMu.Lock()
	t.pending[envelope.MsgID] = replyCh
	t.pendingMu.Unlock()

	frameBytes, err := EncodeFrame(envelope, bodyBytes)
	if err != nil {
		t.removePending(envelope.MsgID)
		return Frame{}, err
	}

	if err := t.writeBinary(ctx, frameBytes); err != nil {
		t.removePending(envelope.MsgID)
		return Frame{}, err
	}

	select {
	case reply := <-replyCh:
		return reply.frame, reply.err
	case <-ctx.Done():
		t.removePending(envelope.MsgID)
		return Frame{}, ctx.Err()
	case <-t.closed:
		t.removePending(envelope.MsgID)
		return Frame{}, t.transportClosedError()
	}
}

func (t *peerTransport) send(ctx context.Context, envelope Envelope, bodyBytes []byte) error {
	frameBytes, err := EncodeFrame(envelope, bodyBytes)
	if err != nil {
		return err
	}
	return t.writeBinary(ctx, frameBytes)
}

func (t *peerTransport) call(ctx context.Context, targetPeerID string, function string, args []any) (Frame, error) {
	bodyBytes, err := EncodeMsgpack(CallBody{Function: function, Args: args})
	if err != nil {
		return Frame{}, &TransportError{Origin: ErrorOriginTransport, Err: err}
	}
	envelope := NewEnvelope(KindCall, t.peerID, targetPeerID)
	return t.request(ctx, envelope, bodyBytes)
}

func (t *peerTransport) queryExists(ctx context.Context, peerID string) (QueryExistsResponseBody, error) {
	bodyBytes, err := EncodeMsgpack(QueryBody{Query: QueryPeerExists, PeerID: peerID})
	if err != nil {
		return QueryExistsResponseBody{}, &TransportError{Origin: ErrorOriginTransport, Err: err}
	}

	frame, err := t.request(ctx, NewEnvelope(KindQuery, t.peerID, RouterPeerID), bodyBytes)
	if err != nil {
		return QueryExistsResponseBody{}, err
	}
	if frame.Envelope.Kind == KindError {
		body, decodeErr := decodeWireErrorBody(frame.BodyBytes)
		if decodeErr != nil {
			return QueryExistsResponseBody{}, decodeErr
		}
		return QueryExistsResponseBody{}, ErrorFromWire(body)
	}

	var response QueryExistsResponseBody
	if err := DecodeMsgpack(frame.BodyBytes, &response); err != nil {
		return QueryExistsResponseBody{}, &RouterError{
			Origin:  ErrorOriginProtocol,
			Code:    "malformed_query_response",
			Message: err.Error(),
		}
	}
	return response, nil
}

func (t *peerTransport) queryGet(ctx context.Context, peerID string) (QueryGetResponseBody, error) {
	bodyBytes, err := EncodeMsgpack(QueryBody{Query: QueryPeerGet, PeerID: peerID})
	if err != nil {
		return QueryGetResponseBody{}, &TransportError{Origin: ErrorOriginTransport, Err: err}
	}

	frame, err := t.request(ctx, NewEnvelope(KindQuery, t.peerID, RouterPeerID), bodyBytes)
	if err != nil {
		return QueryGetResponseBody{}, err
	}
	if frame.Envelope.Kind == KindError {
		body, decodeErr := decodeWireErrorBody(frame.BodyBytes)
		if decodeErr != nil {
			return QueryGetResponseBody{}, decodeErr
		}
		return QueryGetResponseBody{}, ErrorFromWire(body)
	}

	var response QueryGetResponseBody
	if err := DecodeMsgpack(frame.BodyBytes, &response); err != nil {
		return QueryGetResponseBody{}, &RouterError{
			Origin:  ErrorOriginProtocol,
			Code:    "malformed_query_response",
			Message: err.Error(),
		}
	}
	return response, nil
}

func (t *peerTransport) disconnect(ctx context.Context) error {
	bodyBytes := []byte{}
	envelope := NewEnvelope(KindDisconnect, t.peerID, RouterPeerID)
	if err := t.send(ctx, envelope, bodyBytes); err != nil {
		return err
	}
	return nil
}

func (t *peerTransport) readLoop() {
	defer t.closeWithError(nil)

	for {
		msgType, bytes, err := t.conn.Read(context.Background())
		if err != nil {
			t.closeWithError(t.wrapReadError(err))
			return
		}
		if msgType != websocket.MessageBinary {
			t.closeWithError(&RouterError{
				Origin:  ErrorOriginProtocol,
				Code:    "non_binary_message",
				Message: "router must send binary websocket frames",
			})
			return
		}

		frame, err := DecodeFrame(bytes)
		if err != nil {
			t.closeWithError(&RouterError{
				Origin:  ErrorOriginProtocol,
				Code:    "malformed_frame",
				Message: err.Error(),
			})
			return
		}

		if t.deliverPending(frame) {
			continue
		}

		if t.dispatch != nil {
			frameCopy := frame
			go t.dispatch(frameCopy)
		}

		if frame.Envelope.Kind == KindDisconnect {
			t.closeWithError(&TransportError{
				Origin: ErrorOriginTransport,
				Err:    errors.New("peer disconnected"),
			})
			return
		}
	}
}

func (t *peerTransport) deliverPending(frame Frame) bool {
	t.pendingMu.Lock()
	replyCh, ok := t.pending[frame.Envelope.MsgID]
	if ok {
		delete(t.pending, frame.Envelope.MsgID)
	}
	t.pendingMu.Unlock()
	if !ok {
		return false
	}

	var replyErr error
	if frame.Envelope.Kind == KindError {
		body, decodeErr := decodeWireErrorBody(frame.BodyBytes)
		if decodeErr != nil {
			replyErr = decodeErr
		} else {
			replyErr = ErrorFromWire(body)
		}
	}
	replyCh <- transportReply{frame: frame, err: replyErr}
	close(replyCh)
	return true
}

func (t *peerTransport) removePending(msgID string) {
	t.pendingMu.Lock()
	delete(t.pending, msgID)
	t.pendingMu.Unlock()
}

func (t *peerTransport) writeBinary(ctx context.Context, bytes []byte) error {
	if t.isClosed() {
		return t.transportClosedError()
	}

	t.writeMu.Lock()
	defer t.writeMu.Unlock()

	if t.isClosed() {
		return t.transportClosedError()
	}

	if err := t.conn.Write(ctx, websocket.MessageBinary, bytes); err != nil {
		wrapped := t.wrapReadError(err)
		t.closeWithError(wrapped)
		return wrapped
	}
	return nil
}

func (t *peerTransport) closeWithError(err error) {
	t.closedOnce.Do(func() {
		if err != nil {
			t.closeErr.Store(err)
		}
		close(t.closed)
		t.cancelPending(err)
		_ = t.conn.Close(websocket.StatusNormalClosure, "")
	})
}

func (t *peerTransport) cancelPending(err error) {
	if err == nil {
		err = t.transportClosedError()
	}

	t.pendingMu.Lock()
	defer t.pendingMu.Unlock()
	for msgID, replyCh := range t.pending {
		select {
		case replyCh <- transportReply{err: err}:
		default:
		}
		close(replyCh)
		delete(t.pending, msgID)
	}
}

func (t *peerTransport) transportClosedError() error {
	if value := t.closeErr.Load(); value != nil {
		if err, ok := value.(error); ok && err != nil {
			return err
		}
	}
	return &TransportError{Origin: ErrorOriginTransport, Err: errors.New("transport closed")}
}

func (t *peerTransport) wrapReadError(err error) error {
	if err == nil {
		return nil
	}
	return &TransportError{Origin: ErrorOriginTransport, Err: err}
}

func (t *peerTransport) isClosed() bool {
	select {
	case <-t.closed:
		return true
	default:
		return false
	}
}

func decodeWireErrorBody(bodyBytes []byte) (ErrorBody, error) {
	var body ErrorBody
	if err := DecodeMsgpack(bodyBytes, &body); err != nil {
		return ErrorBody{}, &RouterError{
			Origin:  ErrorOriginProtocol,
			Code:    "malformed_error_body",
			Message: err.Error(),
		}
	}
	return body, nil
}

File: php/src/Transport.php
<?php

declare(strict_types=1);

namespace Multifrost;

use WebSocket\Client;
use WebSocket\Configuration;
use WebSocket\Middleware\CloseHandler;
use WebSocket\Middleware\PingResponder;
use WebSocket\Message\Message;
use WebSocket\Exception\ExceptionInterface as WebSocketException;

/**
 * Live WebSocket peer transport.
 * Mirrors golang/transport.go.
 *
 * For callers: send a request, then read responses until msg_id matches.
 * For services: the read loop dispatches incoming calls to a handler.
 */
final class PeerTransport
{
    private Client $client;
    private FrameCodec $codec;
    private bool $closed = false;

    public function __construct(
        public readonly string $peerId,
        public readonly PeerClass $class,
    ) {
        $this->codec = new FrameCodec();
    }

    /**
     * Open WebSocket, send register, wait for ack.
     * Returns the connected PeerTransport.
     */
    /**
     * @param \Closure(Frame): void|null $dispatch unused; kept for API compatibility
     */
    public static function dial(
        string $endpoint,
        string $peerId,
        PeerClass $class,
        ?\Closure $dispatch = null,
        int $timeout = 5,
    ): self {
        $transport = new self($peerId, $class);

        try {
            $config = new Configuration(timeout: $timeout);
            $transport->client = new Client($endpoint, $config);
            $transport->client
                ->addMiddleware(new CloseHandler())
                ->addMiddleware(new PingResponder());
        } catch (\Throwable $e) {
            throw new TransportError(
                \sprintf('websocket dial failed: %s', $e->getMessage()),
                $e,
            );
        }

        $transport->register();

        return $transport;
    }

    /**
     * Send register frame and wait for ack.
     */
    private function register(): void
    {
        $body = ['peer_id' => $this->peerId, 'class' => $this->class->value];
        $bodyBytes = $this->codec->encode($body);
        $envelope = Protocol::newEnvelope(Protocol::KIND_REGISTER, $this->peerId, Protocol::ROUTER_PEER_ID);
        $frameBytes = $this->codec->encodeFrame($envelope, $bodyBytes);

        $this->writeBinary($frameBytes);

        $responseFrame = $this->readOneFrame();

        switch ($responseFrame->envelope->kind) {
            case Protocol::KIND_RESPONSE:
                $ack = $this->codec->decode($responseFrame->bodyBytes);
                if (!\is_array($ack)) {
                    throw new RegistrationError('invalid register ack format');
                }
                if (!($ack['accepted'] ?? false)) {
                    $reason = $ack['reason'] ?? 'router rejected registration';
                    throw new RegistrationError((string) $reason);
                }
                return;

            case Protocol::KIND_ERROR:
                $errBody = $this->decodeErrorBody($responseFrame->bodyBytes);
                throw new RegistrationError($errBody->message);

            default:
                throw new RegistrationError(
                    \sprintf('unexpected register reply kind: %s', $responseFrame->envelope->kind)
                );
        }
    }

    /**
     * Blocking call: send call frame, then read until matching msg_id response/error.
     *
     * @param list<mixed> $args
     */
    public function call(string $targetPeerId, string $function, array $args = []): mixed
    {
        $body = ['function' => $function, 'args' => $args];
        $bodyBytes = $this->codec->encode($body);
        $envelope = Protocol::newEnvelope(Protocol::KIND_CALL, $this->peerId, $targetPeerId);
        $msgId = $envelope->msgId;

        $frameBytes = $this->codec->encodeFrame($envelope, $bodyBytes);
        $this->writeBinary($frameBytes);

        // Read until we get a response/error with matching msg_id
        while (true) {
            $responseFrame = $this->readOneFrame();

            if ($responseFrame->envelope->msgId !== $msgId) {
                // Unsolicited frame — dispatch if service mode, otherwise skip
                continue;
            }

            switch ($responseFrame->envelope->kind) {
                case Protocol::KIND_RESPONSE:
                    $decoded = $this->codec->decode($responseFrame->bodyBytes);
                    if (\is_array($decoded) && \array_key_exists('result', $decoded)) {
                        return $decoded['result'];
                    }
                    return $decoded;

                case Protocol::KIND_ERROR:
                    $errBody = $this->decodeErrorBody($responseFrame->bodyBytes);
                    throw errorFromWire($errBody);

                default:
                    throw new RouterError(
                        'unexpected_response_kind',
                        \sprintf('unexpected frame kind in call response: %s', $responseFrame->envelope->kind),
                    );
            }
        }
    }

    /**
     * Query whether a peer exists.
     */
    public function queryExists(string $peerId): QueryExistsResponseBody
    {
        $body = ['query' => Protocol::QUERY_PEER_EXISTS, 'peer_id' => $peerId];
        $bodyBytes = $this->codec->encode($body);
        $envelope = Protocol::newEnvelope(Protocol::KIND_QUERY, $this->peerId, Protocol::ROUTER_PEER_ID);
        $msgId = $envelope->msgId;

        $frameBytes = $this->codec->encodeFrame($envelope, $bodyBytes);
        $this->writeBinary($frameBytes);

        $responseFrame = $this->readOneFrame();

        // The response msg_id should match, but handle mismatch gracefully
        // (query is simple request/response, router shouldn't interleave)

        if ($responseFrame->envelope->kind === Protocol::KIND_ERROR) {
            $errBody = $this->decodeErrorBody($responseFrame->bodyBytes);
            throw errorFromWire($errBody);
        }

        /** @var array<string, mixed> $decoded */
        $decoded = $this->codec->decode($responseFrame->bodyBytes);
        if (!\is_array($decoded)) { // @phpstan-ignore function.alreadyNarrowedType
            throw new RouterError('malformed_query_response', 'query response must be an object');
        }

        return new QueryExistsResponseBody(
            peerId: (string) ($decoded['peer_id'] ?? ''),
            exists: (bool) ($decoded['exists'] ?? false),
            class: isset($decoded['class']) ? PeerClass::tryFrom((string) $decoded['class']) : null,
            connected: (bool) ($decoded['connected'] ?? false),
        );
    }

    /**
     * Query full peer details.
     */
    public function queryGet(string $peerId): QueryGetResponseBody
    {
        $body = ['query' => Protocol::QUERY_PEER_GET, 'peer_id' => $peerId];
        $bodyBytes = $this->codec->encode($body);
        $envelope = Protocol::newEnvelope(Protocol::KIND_QUERY, $this->peerId, Protocol::ROUTER_PEER_ID);
        $msgId = $envelope->msgId;

        $frameBytes = $this->codec->encodeFrame($envelope, $bodyBytes);
        $this->writeBinary($frameBytes);

        $responseFrame = $this->readOneFrame();

        if ($responseFrame->envelope->kind === Protocol::KIND_ERROR) {
            $errBody = $this->decodeErrorBody($responseFrame->bodyBytes);
            throw errorFromWire($errBody);
        }

        /** @var array<string, mixed> $decoded */
        $decoded = $this->codec->decode($responseFrame->bodyBytes);
        if (!\is_array($decoded)) { // @phpstan-ignore function.alreadyNarrowedType
            throw new RouterError('malformed_query_response', 'query response must be an object');
        }

        return new QueryGetResponseBody(
            peerId: (string) ($decoded['peer_id'] ?? ''),
            exists: (bool) ($decoded['exists'] ?? false),
            class: isset($decoded['class']) ? PeerClass::tryFrom((string) $decoded['class']) : null,
            connected: (bool) ($decoded['connected'] ?? false),
        );
    }

    /**
     * Graceful disconnect.
     */
    public function disconnect(): void
    {
        if ($this->closed) {
            return;
        }
        $envelope = Protocol::newEnvelope(Protocol::KIND_DISCONNECT, $this->peerId, Protocol::ROUTER_PEER_ID);
        $frameBytes = $this->codec->encodeFrame($envelope, '');
        try {
            $this->writeBinary($frameBytes);
        } catch (\Throwable) {
            // Best effort
        }
        $this->close();
    }

    /**
     * Close the WebSocket connection.
     */
    public function close(): void
    {
        if ($this->closed) {
            return;
        }
        $this->closed = true;
        try {
            $this->client->close();
        } catch (\Throwable) {
            // Best effort
        }
    }

    /**
     * Service read loop: receive frames and dispatch to handler.
     * Blocks until connection closes or dispatch callback throws.
     */
    public function serve(\Closure $handler): void
    {
        while (!$this->closed) {
            try {
                $frame = $this->readOneFrame();
            } catch (WebSocketException) {
                $this->closed = true;
                break;
            } catch (TransportError) {
                $this->closed = true;
                break;
            }

            switch ($frame->envelope->kind) {
                case Protocol::KIND_CALL:
                    $handler($frame);
                    break;
                case Protocol::KIND_DISCONNECT:
                    $this->close();
                    break 2;
                case Protocol::KIND_HEARTBEAT:
                    // Ignore
                    break;
                default:
                    // Ignore unexpected kinds in service mode
                    break;
            }
        }
    }

    /**
     * Send a prepared frame (envelope + body bytes) directly.
     */
    public function send(Envelope $envelope, string $bodyBytes): void
    {
        $frameBytes = $this->codec->encodeFrame($envelope, $bodyBytes);
        $this->writeBinary($frameBytes);
    }

    public function isClosed(): bool
    {
        return $this->closed;
    }

    // ── Private helpers ──────────────────────────────────────────

    /**
     * Write raw binary bytes to the WebSocket.
     */
    private function writeBinary(string $bytes): void
    {
        if ($this->closed) {
            throw new TransportError('transport closed');
        }
        try {
            $this->client->binary($bytes);
        } catch (\Throwable $e) {
            $this->close();
            throw new TransportError($e->getMessage(), $e);
        }
    }

    /**
     * Read one binary frame from the WebSocket, blocking.
     */
    private function readOneFrame(): Frame
    {
        try {
            $message = $this->client->receive();
            $content = $message->getContent();
            return $this->codec->decodeFrame($content);
        } catch (WebSocketException $e) {
            $this->closed = true;
            throw new TransportError(
                \sprintf('websocket connection error: %s', $e->getMessage()),
                $e,
            );
        } catch (\InvalidArgumentException $e) {
            throw new TransportError(
                \sprintf('frame decode error: %s', $e->getMessage()),
                $e,
            );
        }
    }

    /**
     * Decode an error body from raw bytes.
     */
    private function decodeErrorBody(string $bodyBytes): ErrorBody
    {
        $decoded = $this->codec->decode($bodyBytes);
        if (!\is_array($decoded)) {
            throw new RouterError('malformed_error_body', 'error body must be an object');
        }

        return new ErrorBody(
            code: (string) ($decoded['code'] ?? 'unknown'),
            message: (string) ($decoded['message'] ?? ''),
            kind: (string) ($decoded['kind'] ?? ErrorOrigin::Router->value),
            stack: isset($decoded['stack']) ? (string) $decoded['stack'] : null,
            details: $decoded['details'] ?? null,
        );
    }
}
